diff -uNr i3-4.8/include/atoms.xmacro i3-4.8p/include/atoms.xmacro
--- i3-4.8/include/atoms.xmacro	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/include/atoms.xmacro	2016-05-13 09:28:23.231844563 +0200
@@ -34,3 +34,5 @@
 xmacro(_NET_REQUEST_FRAME_EXTENTS)
 xmacro(_NET_FRAME_EXTENTS)
 xmacro(_MOTIF_WM_HINTS)
+xmacro(_QUBES_VMNAME)
+xmacro(_QUBES_LABEL)
diff -uNr i3-4.8/include/data.h i3-4.8p/include/data.h
--- i3-4.8/include/data.h	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/include/data.h	2016-05-13 09:28:23.232844563 +0200
@@ -348,6 +348,12 @@
     /** The name of the window. */
     i3String *name;
 
+    /** The name of the qubes vm. */
+    i3String *qubes_vmname;
+
+    /** The qubes label. */
+    int qubes_label;
+
     /** The WM_WINDOW_ROLE of this window (for example, the pidgin buddy window
      * sets "buddy list"). Useful to match specific windows in assignments or
      * for_window. */
diff -uNr i3-4.8/include/window.h i3-4.8p/include/window.h
--- i3-4.8/include/window.h	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/include/window.h	2016-05-13 09:28:23.233844563 +0200
@@ -33,6 +33,20 @@
 void window_update_name_legacy(i3Window *win, xcb_get_property_reply_t *prop, bool before_mgmt);
 
 /**
+ * Updates the qubes vmname by using _QUBES_VMNAME (encoded in UTF-8) for the given
+ * window.
+ *
+ */
+void window_update_qubes_vmname(i3Window *win, xcb_get_property_reply_t *prop, bool before_mgmt);
+
+/**
+ * Updates the qubes label by using _QUBES_LABEL (encoded in UTF-8) for the given
+ * window.
+ *
+ */
+void window_update_qubes_label(i3Window *win, xcb_get_property_reply_t *prop, bool before_mgmt);
+
+/**
  * Updates the CLIENT_LEADER (logical parent window).
  *
  */
diff -uNr i3-4.8/src/commands.c i3-4.8p/src/commands.c
--- i3-4.8/src/commands.c	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/src/commands.c	2016-05-13 09:34:54.964844563 +0200
@@ -836,23 +836,17 @@
         if (strcmp(border_style_str, "toggle") == 0) {
             border_style++;
             border_style %= 3;
-            if (border_style == BS_NORMAL)
-                tmp_border_width = 2;
-            else if (border_style == BS_NONE)
-                tmp_border_width = 0;
-            else if (border_style == BS_PIXEL)
-                tmp_border_width = 1;
+            tmp_border_width = 3;
         } else {
             if (strcmp(border_style_str, "normal") == 0)
                 border_style = BS_NORMAL;
             else if (strcmp(border_style_str, "pixel") == 0)
                 border_style = BS_PIXEL;
-            else if (strcmp(border_style_str, "1pixel") == 0) {
+            else if (strcmp(border_style_str, "1pixel") == 0 ||
+                     strcmp(border_style_str, "none") == 0) {
                 border_style = BS_PIXEL;
-                tmp_border_width = 1;
-            } else if (strcmp(border_style_str, "none") == 0)
-                border_style = BS_NONE;
-            else {
+                tmp_border_width = 3;
+            } else {
                 ELOG("BUG: called with border_style=%s\n", border_style_str);
                 ysuccess(false);
                 return;
diff -uNr i3-4.8/src/config.c i3-4.8p/src/config.c
--- i3-4.8/src/config.c	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/src/config.c	2016-05-13 09:28:23.237844563 +0200
@@ -264,6 +264,12 @@
 
     parse_configuration(override_configpath);
 
+    /* redefine defaults, to overwrite user settings easily */
+    config.default_border = BS_NORMAL;
+    config.default_floating_border = BS_NORMAL;
+    config.default_border_width = logical_px(2);
+    config.default_floating_border_width = logical_px(2);
+
     if (reload) {
         translate_keysyms();
         grab_all_keys(conn, false);
diff -uNr i3-4.8/src/manage.c i3-4.8p/src/manage.c
--- i3-4.8/src/manage.c	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/src/manage.c	2016-05-13 09:28:23.239844563 +0200
@@ -87,7 +87,7 @@
     xcb_get_window_attributes_reply_t *attr = NULL;
 
     xcb_get_property_cookie_t wm_type_cookie, strut_cookie, state_cookie,
-        utf8_title_cookie, title_cookie,
+        utf8_title_cookie, title_cookie, qubes_vmname_cookie, qubes_label_cookie,
         class_cookie, leader_cookie, transient_cookie,
         role_cookie, startup_id_cookie, wm_hints_cookie,
         wm_normal_hints_cookie, motif_wm_hints_cookie;
@@ -152,6 +152,8 @@
     strut_cookie = GET_PROPERTY(A__NET_WM_STRUT_PARTIAL, UINT32_MAX);
     state_cookie = GET_PROPERTY(A__NET_WM_STATE, UINT32_MAX);
     utf8_title_cookie = GET_PROPERTY(A__NET_WM_NAME, 128);
+    qubes_vmname_cookie = GET_PROPERTY(A__QUBES_VMNAME, 128);
+    qubes_label_cookie = GET_PROPERTY(A__QUBES_LABEL, UINT32_MAX);
     leader_cookie = GET_PROPERTY(A_WM_CLIENT_LEADER, UINT32_MAX);
     transient_cookie = GET_PROPERTY(XCB_ATOM_WM_TRANSIENT_FOR, UINT32_MAX);
     title_cookie = GET_PROPERTY(XCB_ATOM_WM_NAME, 128);
@@ -188,6 +190,8 @@
     window_update_class(cwindow, xcb_get_property_reply(conn, class_cookie, NULL), true);
     window_update_name_legacy(cwindow, xcb_get_property_reply(conn, title_cookie, NULL), true);
     window_update_name(cwindow, xcb_get_property_reply(conn, utf8_title_cookie, NULL), true);
+    window_update_qubes_vmname(cwindow, xcb_get_property_reply(conn, qubes_vmname_cookie, NULL), true);
+    window_update_qubes_label(cwindow, xcb_get_property_reply(conn, qubes_label_cookie, NULL), true);
     window_update_leader(cwindow, xcb_get_property_reply(conn, leader_cookie, NULL));
     window_update_transient_for(cwindow, xcb_get_property_reply(conn, transient_cookie, NULL));
     window_update_strut_partial(cwindow, xcb_get_property_reply(conn, strut_cookie, NULL));
diff -uNr i3-4.8/src/window.c i3-4.8p/src/window.c
--- i3-4.8/src/window.c	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/src/window.c	2016-05-13 09:28:23.240844563 +0200
@@ -120,6 +120,57 @@
 }
 
 /*
+ * Updates the qubes vmname by using _QUBES_VMNAME (encoded in UTF-8) for the given
+ * window.
+ *
+ */
+void window_update_qubes_vmname(i3Window *win, xcb_get_property_reply_t *prop, bool before_mgmt) {
+    if (prop == NULL || xcb_get_property_value_length(prop) == 0) {
+        win->qubes_vmname = i3string_from_utf8("dom0");
+        FREE(prop);
+        return;
+    }
+
+    i3string_free(win->qubes_vmname);
+    win->qubes_vmname = i3string_from_utf8_with_length(xcb_get_property_value(prop),
+                                                       xcb_get_property_value_length(prop));
+    LOG("_QUBES_VMNAME set to \"%s\"\n", i3string_as_utf8(win->qubes_vmname));
+
+    if (before_mgmt) {
+        free(prop);
+        return;
+    }
+
+    run_assignments(win);
+
+    free(prop);
+}
+
+/*
+ * Updates the qubes label by using _QUBES_LABEL (encoded in UTF-8) for the given
+ * window.
+ *
+ */
+void window_update_qubes_label(i3Window *win, xcb_get_property_reply_t *prop, bool before_mgmt) {
+    if (prop == NULL || xcb_get_property_value_length(prop) == 0) {
+        win->qubes_label = 0;
+        FREE(prop);
+        return;
+    }
+
+    win->qubes_label = *(int*) xcb_get_property_value(prop);
+
+    if (before_mgmt) {
+        free(prop);
+        return;
+    }
+
+    run_assignments(win);
+
+    free(prop);
+}
+
+/*
  * Updates the CLIENT_LEADER (logical parent window).
  *
  */
diff -uNr i3-4.8/src/x.c i3-4.8p/src/x.c
--- i3-4.8/src/x.c	2014-06-15 19:12:43.000000000 +0200
+++ i3-4.8p/src/x.c	2016-05-13 09:33:27.623844563 +0200
@@ -348,6 +348,50 @@
     else
         p->color = &config.client.unfocused;
 
+    struct Window *win = con->window;
+    if (win != NULL) {
+        bool in_focus = con == focused || con_inside_focused(con);
+        DLOG("con->qubes_label is %d\n", win->qubes_label);
+        switch (win->qubes_label) {
+        case 0: // dom0
+            p->color->text = get_colorpixel("#ffffff");
+            p->color->background = get_colorpixel(in_focus ? "#474747" : "#303030");
+            break;
+        case 1: // red
+            p->color->text = get_colorpixel("#000000");
+            p->color->background = get_colorpixel(in_focus ? "#ff0000" : "#ba0000");
+            break;
+        case 2: // orange
+            p->color->text = get_colorpixel("#0059FF");
+            p->color->background = get_colorpixel(in_focus ? "#ffa500" : "#b87700");
+            break;
+        case 3: // yellow
+            p->color->text = get_colorpixel("#000000");
+            p->color->background = get_colorpixel(in_focus ? "#ffff00" : "#baba00");
+            break;
+        case 4: // green
+            p->color->text = get_colorpixel("#ffffff");
+            p->color->background = get_colorpixel(in_focus ? "#347235" : "#214822");
+            break;
+        case 5: // gray
+            p->color->text = get_colorpixel("#ffffff");
+            p->color->background = get_colorpixel(in_focus ? "#777775" : "#5d5d5b");
+            break;
+        case 6: // blue
+            p->color->text = get_colorpixel("#000000");
+            p->color->background = get_colorpixel(in_focus ? "#1e5ef3" : "#133b98");
+            break;
+        case 7: // purple
+            p->color->text = get_colorpixel("#ffffff");
+            p->color->background = get_colorpixel(in_focus ? "#9a219b" : "#621562");
+            break;
+        case 8: // black
+            p->color->text = get_colorpixel("#ffffff");
+            p->color->background = get_colorpixel(in_focus ? "#2f2f2f" : "#000000");
+            break;
+        }
+    }
+
     p->border_style = con_border_style(con);
 
     Rect *r = &(con->rect);
@@ -492,7 +536,6 @@
     set_font_colors(parent->pm_gc, p->color->text, p->color->background);
     int text_offset_y = (con->deco_rect.height - config.font.height) / 2;
 
-    struct Window *win = con->window;
     if (win == NULL) {
         /* we have a split container which gets a representation
          * of its children as title
@@ -531,10 +574,16 @@
     //DLOG("indent_level = %d, indent_mult = %d\n", indent_level, indent_mult);
     int indent_px = (indent_level * 5) * indent_mult;
 
-    draw_text(win->name,
-              parent->pixmap, parent->pm_gc,
-              con->deco_rect.x + 2 + indent_px, con->deco_rect.y + text_offset_y,
-              con->deco_rect.width - 2 - indent_px);
+    /* set window title, include qubes vmname */
+    char *vmname_and_name;
+    sasprintf(&vmname_and_name, "[%s] %s", i3string_as_utf8(win->qubes_vmname), i3string_as_utf8(win->name));
+
+    draw_text_ascii(vmname_and_name,
+                    parent->pixmap, parent->pm_gc,
+                    con->deco_rect.x + 2 + indent_px, con->deco_rect.y + text_offset_y,
+                    con->deco_rect.width - 2 - indent_px);
+
+    free(vmname_and_name);
 
 after_title:
     /* Since we don’t clip the text at all, it might in some cases be painted
