From 9dce0818378feefe0aa7844f51f48e03a01607dc Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <michael@stapelberg.de>
Date: Fri, 01 Jan 2010 15:19:42 +0100
Subject: Bugfix: Don’t put dock clients into floating mode (Thanks xeen)

---
diff --git a/src/floating.c b/src/floating.c
index 4177b6e..07bc331 100644
--- i3-3.d-bf1.orig/src/floating.c
+++ i3-3.d-bf1/src/floating.c
@@ -41,6 +41,11 @@ void toggle_floating_mode(xcb_connection_t *conn, Client *client, bool automatic
         Container *con = client->container;
         i3Font *font = load_font(conn, config.font);
 
+        if (client->dock) {
+                DLOG("Not putting dock client into floating mode\n");
+                return;
+        }
+
         if (con == NULL) {
                 LOG("This client is already in floating (container == NULL), re-inserting\n");
                 Client *next_tiling;
--
cgit v0.8.3
