From 8adce413f5f92bd8e4c485ee61ff6b8448cd2058 Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <michael@stapelberg.de>
Date: Fri, 01 Jan 2010 16:30:27 +0100
Subject: Remove superfluous #include <assert.h> (Thanks badboy)

---
diff --git a/src/mainx.c b/src/mainx.c
index 322e51e..d847d6f 100644
--- i3-3.d-bf1.orig/src/mainx.c
+++ i3-3.d-bf1/src/mainx.c
@@ -9,7 +9,6 @@
  *
  */
 #include <stdio.h>
-#include <assert.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/types.h>
--
cgit v0.8.3
