From f399c3ef9d0db8c95113b242e0c69498cf077669 Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <michael@stapelberg.de>
Date: Fri, 22 Jan 2010 23:34:29 +0100
Subject: Bugfix: Don’t leak IPC socket to launched processes

---
diff --git a/src/ipc.c b/src/ipc.c
index 0bef2ea..c048cdf 100644
--- i3-3.d-bf1.orig/src/ipc.c
+++ i3-3.d-bf1/src/ipc.c
@@ -228,6 +228,8 @@ int ipc_create_socket(const char *filename) {
                 return -1;
         }
 
+        (void)fcntl(sockfd, F_SETFD, FD_CLOEXEC);
+
         struct sockaddr_un addr;
         memset(&addr, 0, sizeof(struct sockaddr_un));
         addr.sun_family = AF_LOCAL;
--
cgit v0.8.3
