From 1dcb4a39fd676441c30ce5614c2ae6082114837d Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <michael@stapelberg.de>
Date: Thu, 07 Jan 2010 18:24:46 +0100
Subject: Use LOG instead of DLOG (next branch feature only)

---
diff --git a/src/xcb.c b/src/xcb.c
index 98b6bde..ee75ca2 100644
--- i3-3.d-bf1.orig/src/xcb.c
+++ i3-3.d-bf1/src/xcb.c
@@ -359,7 +359,7 @@ int predict_text_width(xcb_connection_t *conn, const char *font_pattern, char *t
 
         /* If no per-char info is available for this font, we use the default */
         if (xcb_query_font_char_infos_length(font_info) == 0) {
-                DLOG("Falling back on default char_width of %d pixels\n", font_info->max_bounds.character_width);
+                LOG("Falling back on default char_width of %d pixels\n", font_info->max_bounds.character_width);
                 return (font_info->max_bounds.character_width * length);
         }
 
--
cgit v0.8.3
