From 4c87170494cbdd680cb4b128dd60fcdaac995609 Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <michael@stapelberg.de>
Date: Fri, 01 Jan 2010 16:29:11 +0100
Subject: We don’t have DLOG yet in master branch (Thanks ccount)

---
diff --git a/src/floating.c b/src/floating.c
index 07bc331..9302b02 100644
--- i3-3.d-bf1.orig/src/floating.c
+++ i3-3.d-bf1/src/floating.c
@@ -42,7 +42,7 @@ void toggle_floating_mode(xcb_connection_t *conn, Client *client, bool automatic
         i3Font *font = load_font(conn, config.font);
 
         if (client->dock) {
-                DLOG("Not putting dock client into floating mode\n");
+                LOG("Not putting dock client into floating mode\n");
                 return;
         }
 
--
cgit v0.8.3
